src/paths/floyd_warshall.c

```diff
@@ -140,3 +144,131 @@ igraph_error_t igraph_distances_floyd_warshall(
 
     return IGRAPH_SUCCESS;
 }
+
+igraph_error_t igraph_distances_floyd_warshall_tree_speedup(
+        const igraph_t *graph, igraph_matrix_t *res,
+        const igraph_vector_t *weights, igraph_neimode_t mode) {
+
+    igraph_integer_t no_of_nodes = igraph_vcount(graph);
+    igraph_integer_t no_of_edges = igraph_ecount(graph);
+    igraph_bool_t in = false, out = false;
+
+    if (weights && igraph_vector_size(weights) != no_of_edges) {
+        IGRAPH_ERROR("Invalid weight vector length.", IGRAPH_EINVAL);
+    }
+
+    if (! igraph_is_directed(graph)) {
+        mode = IGRAPH_ALL;
+    }
+
+    switch (mode) {
+    case IGRAPH_ALL:
+        in = out = true;
+        break;
+    case IGRAPH_OUT:
+        out = true;
+        break;
+    case IGRAPH_IN:
+        in = true;
+        break;
+    default:
+        IGRAPH_ERROR("Invalid mode.", IGRAPH_EINVAL);
+    }
+
+    if (weights && igraph_vector_is_any_nan(weights)) {
+        IGRAPH_ERROR("Weight vector must not contain NaN values.", IGRAPH_EINVAL);
+    }
+
+    IGRAPH_CHECK(igraph_matrix_resize(res, no_of_nodes, no_of_nodes));
+    igraph_matrix_fill(res, IGRAPH_INFINITY);
+
+    for (igraph_integer_t v=0; v < no_of_nodes; v++) {
+        MATRIX(*res, v, v) = 0;
+    }
+
+    for (igraph_integer_t e=0; e < no_of_edges; e++) {
+        igraph_integer_t from = IGRAPH_FROM(graph, e);
+        igraph_integer_t to = IGRAPH_TO(graph, e);
+        igraph_real_t w = weights ? VECTOR(*weights)[e] : 1;
+
+        if (w < 0) {
+            if (mode == IGRAPH_ALL) {
+                IGRAPH_ERRORF("Negative edge weight (%g) found in undirected graph "
+                              "while calculating distances with Floyd-Warshall.",
+                              IGRAPH_ENEGLOOP, w);
+            } else if (to == from) {
+                IGRAPH_ERRORF("Self-loop with negative weight (%g) found "
+                              "while calculating distances with Floyd-Warshall.",
+                              IGRAPH_ENEGLOOP, w);
+            }
+        }
+
+        if (out && MATRIX(*res, from, to) > w) MATRIX(*res, from, to) = w;
+        if (in  && MATRIX(*res, to, from) > w) MATRIX(*res, to, from) = w;
+    }
+
+    igraph_t out_k;
+    igraph_vector_int_t parents;
+    igraph_matrix_int_t predecessors;
+    igraph_stack_int_t stack;
+    IGRAPH_CHECK(igraph_stack_int_init(&stack, no_of_nodes));
+    IGRAPH_CHECK(igraph_matrix_int_init(&predecessors, no_of_nodes, no_of_nodes));
```

Comment:
```suggestion
    IGRAPH_CHECK(igraph_stack_int_init(&stack, no_of_nodes));
    IGRAPH_FINALLY(igraph_stack_int_destroy, &stack);
    IGRAPH_CHECK(igraph_matrix_int_init(&predecessors, no_of_nodes, no_of_nodes));
    IGRAPH_FINALLY(igraph_matrix_int_destroy, &predecessors);
```

This should fix the memory leak (untested).

I am posting this just for illustration. `IGRAPH_FINALLY` registers a data structure for destruction in case of error. `IGRAPH_FINALLY_CLEAN(n)` discards the last `n` such registrations.

There are often macros that perform both initialization and adding the to the "finally stack". These should be preferred instead of an explicit `IGRAPH_FINALLY` when they are available, just to keep the code shorter. An example of such a macro is `IGRAPH_VECTOR_INT_INIT_FINALLY`. I do not recall immediately if these are available for matrices, but they are certainly available for stacks.